pkg/router/controller/ingress.go

```diff
@@ -0,0 +1,550 @@
+		glog.V(4).Infof("The default backend defined for ingress %v will be ignored.  Default backends are not compatible with the OpenShift router.", key)
+	}
+
+	for _, rule := range ingress.Spec.Rules {
+		for _, path := range rule.HTTP.Paths {
+			name := generateRouteName(ingress.Name, rule.Host, path.Path)
+			if _, ok := routeMap[name]; ok {
+				glog.V(4).Infof("Ingress %s has more than one rule for host '%s' and path '%s'",
+					ingress.Name, rule.Host, path.Path)
+				continue
+			}
+
+			route := &routeapi.Route{
+				ObjectMeta: kapi.ObjectMeta{
+					Name: name,
+					// Reuse the values from ingress
+					Namespace:         ingress.Namespace,
+					CreationTimestamp: ingress.CreationTimestamp,
+					// Append a hash of the name to the ingress uid to ensure both uniqueness
+					// and consistent sorting.
+					UID: types.UID(fmt.Sprintf("%s-%x", ingress.UID, md5.Sum([]byte(name)))),
+				},
+				Spec: routeapi.RouteSpec{
+					Host: rule.Host,
+					Path: path.Path,
+					To: routeapi.RouteTargetReference{
+						Name: path.Backend.ServiceName,
+					},
+					Port: &routeapi.RoutePort{
+						TargetPort: path.Backend.ServicePort,
+					},
+				},
+			}
+
+			// The router depends on defaults being set on resource creation.  Manually
+			// set the same defaults for a generated route to ensure compatibility.
+			//
+			// TODO Consider round tripping the route through the api
+			// conversion to simplify default maintenance.
+
+			// From SetDefaults_RouteSpec
+			route.Spec.WildcardPolicy = routeapi.WildcardPolicyNone
+
+			// From SetDefaults_RouteTargetReference
+			route.Spec.To.Kind = "Service"
+			route.Spec.To.Weight = new(int32)
+			*route.Spec.To.Weight = 100
+
+			routeMap[name] = route
+		}
+	}
+
+	return routeMap
+}
+
+// matchHost checks whether a pattern (which can represent a wildcard
+// domain of the form *.[subdomain]) matches the given host.
+func matchesHost(pattern, host string) bool {
+	if len(pattern) == 0 || len(host) == 0 {
+		return false
+	}
+
+	patternParts := strings.Split(pattern, ".")
+	hostParts := strings.Split(host, ".")
+
+	if len(patternParts) != len(hostParts) {
+		return false
+	}
+
+	for i, patternPart := range patternParts {
+		if i == 0 && patternPart == "*" {
+			continue
+		}
+		if patternPart != hostParts[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func getKey(namespace, name string) string {
+	return fmt.Sprintf("%v/%v", namespace, name)
+}
+
+// generateRouteName returns a stable route name for an ingress name,
+// host and path.  It's fine if the host or path happen to be empty.
+func generateRouteName(name, host, path string) string {
+	// Routes generated from ingress rules contain '/' to prevent name
+	// clashes with user-defined routes.  '/' is not permitted in a resource
+	// name submitted to the api, but generated routes are only intended to
+	// be used in the context of the router controller.
+	//
+	// TODO should the path be hashed or otherwise encoded for compatibility?
```

Comment:
The name of the route (with ``-`` and ``/`` replaced with ``_``) is used in haproxy configuration.  I'm assuming there are other characters that would be legal in a path that would not be legal in haproxy configuration (e.g. ``?`` ``&`` and other characters related to a path with query parameters).